README
OSD Picture in Picture plugin for the Video Disk Recorder
---------------------------------------------------------

Copyright (C) 2010        Mitchm at vdrportal.de
Copyright (C) 2004 - 2008 Andreas Regel <andreas.regel@powarman.de>
Copyright (C) 2004 - 2005 Sascha Volkenandt <sascha@akv-soft.de>

Project's homepage: http://projects.vdr-developer.org/projects/show/plg-osdpip

Old homepages:      http://linux.kompiliert.net/index.php?view=osdpip
                    http://www.powarman.de/files/osdpip

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

See the file COPYING for the full license information.


Requirements:
-------------

- vdr 1.7.27 or higher
- libavcodec from ffmpeg (http://ffmpeg.sourceforge.net) version 0.4.8 or
  higher


Description:
------------

OSD Picture-in-Picture is a PlugIn that displays the current channel in a
small box on the screen (default upper right corner). You can switch up and
down now, watching the progress of the previous channel in the box. Quality is
not too good yet, and only I-Frames are displayed.
The plugin supports five modes:
 - greyscaled (16 shades of grey)
 - greyscaled (256 shades of grey)
 - 256 colors with fixed palette
 - 256 colors with dithering
 - 128 colors with variable palette
They all work with an ordinary vdr installation.
The plugin has the possibility to choose the PiP size (in the setup menu from
predefined sizes) and position (by moving it around using the cursor keys) as
well as crop dimensions.

ATTENTION: To have the position saved, you have to enter the plugin's setup
menu and hit the ok button.

The plugin lets you choose the MPEG frames to decode and display (in the setup
menu). For slower machines it is further possible to set an additional frame
dropping.

The plugin is able to show some channel and schedule info in a second window
when switching channels. In setup menu you can change it's position, size and
the amount of information shown.


Installation:
-------------

Let's say osdpip's version is 0.0.9 and vdr's version is 1.4.0. If you
use anything else please exchange the version numbers appropriately (this
way I don't have to update this section all the times;) ).

cd vdr-1.4.0/PLUGINS/src
tar xvfz vdr-osdpip-0.0.9.tgz
ln -s osdpip-0.0.9 osdpip
cd ../..

If you have ffmpeg installed, "make plugins" should do just fine now, but if
you have it next to your vdr folder (e.g. as the DXR3 plugin needs it), you
have to compile with "make FFMPEG_STATIC=1 plugins". Ffmpeg must then be
present in a folder (or symlink) called "ffmpeg".

In recent FFMPEG versions the functions img_convert and img_replace where
replaced by swscale functions. This are used by default. If you don't want to
use swscale functions for scaling and color format conversions, you will have
to add WITHOUT_SWSCALE=1 to your Make.config or uncomment this line in
vdr-osdpip's Makefile. Then deprecated img_convert and img_resample will be
used instead.

If you have a recent FFMPEG version that has a changed structure of its header
files you will have to add WITH_NEW_FFMPEG_HEADERS=1 to your Make.config or
uncomment this line in vdr-osdpip's Makefile.

Controls:
---------

Normal Mode:
Up/Down      Switch channel up/down
Left/Right   Select channel group (only when info window enabled)
1-9          Select channel (only when info window enabled)
Red          Swap PiP channel with currently viewed channel
Green        Switch to move mode
Yellow       Switch to zapping mode
Ok           Show/hide channel info window
0            Switch back to PiP channel and exit PiP
Back         Exit PiP without switching back

Move Mode:
Cursor keys  Move PiP window around
Green        Switch to normal mode
0            Switch back to PiP channel and exit PiP
Back         Exit PiP without switching back

Zapping Mode:
Up/Down      Switch PiP channel up/down
Red          Swap PiP channel with currently viewed channel
Green        Switch to move mode
Yellow       Switch to normal mode
0            Switch back to PiP channel and exit PiP
Back         Exit PiP without switching back



HISTORY
VDR Plugin 'osdpip' Revision History
------------------------------------

2012-05-16: Version 0.1.2
- Adapted for VDR 1.7.27 (Closes #981, #964, #902)
- Dropped VDR 1.6 support
- Added some new resolutions:
  4:3 - 450x360, 600x480
  16:9 - 320x180, 480x270, 640x360

2012-02-12: Version 0.1.1

- Fixed compile error with newer ffmpeg versions >= svn 20100426 (closes #345)
- Replaced deprected ffmpeg calls (Patch provided by mnival) (closes #690)
- Updated french translation (Patch provided by mnival) (closes #691)
- Avoid segfault at stop replay with keystroke '0' (Patch provided by
  Andreas Brachold) (closes #592)

2010-03-28: Version 0.1.0

- Moved to http://projects.vdr-developer.org/projects/show/plg-osdpip
  as community maintained project using a GIT repository
- Committed old releases to the GIT repository
- Updated project homepage url in README
- Updated FSF address in the license information
- Added support for VDR >= 1.7.13 (Make.global)
- Use pkg-config to find compilation flags, libs etc for ffmpeg
- Fixed ffmpeg compatibility issues with newer ffmpeg versions
- Improved copyright and license information in the header of all source files

2010-01-01: Version 0.0.10a
- added support for VDR 1.7.x
- added PIP while replaying
- added italian translations (thanks to Diego Pierotto)

2008-05-03: Version 0.0.10 (written by Andreas Regel)
- support swscale functions of recent FFMPEG versions. Have a look at
  README to see how to deactivate it for older FFPMEG versions.
- support changed header file structure of recent FFMPEG versions. Have
  a look at README to see how to activate this.
- added zapping through PiP channel based on a patch by pinky666 from
  vdr-portal. You can activate it via the green button.

2008-04-27: Version 0.0.9 (written by Andreas Regel)
- dropped support for VDR version lower than 1.3.47.
- Now using palette replacemen introduced in VDR 1.4.27, so palette
  setting workaround is not needed any longer. This simplifies code a
  bit.
- added support for VDR 1.6. Now 1.4 and 1.6 are supported (tested with
  1.4.7 and 1.6.0)
- added gettext localization support introduced with VDR 1.5.7.
- Do not decode B frames when dropping them.
- Use cCondWait::SleepMs instead of usleep.
- Fixed automatic closing of OSD after some minutes.
- Fixed showing wrong channel information.
- Added missing inclusion of <vdr/device.h>.
- Added new colour reduction mode using 256 dithered colours (thanks to
  Martin Wache).
- Added french translations (thanks to micky979 at vdrportal.de)


2005-05-15: Version 0.0.8 (written by Andreas Regel)
- Added Navigation through channel in a way like with ordinary VDR. Use
  the Left/Right keys to select channel groups, OK to choose one or the
  number keys to directly enter a channel number. All this only works
  with enabled info window.
- Extended the info window to contain more information. The amount of
  information shown can be chosen through the setup menu. This directly
  affects the height of the info window.
- Using the info window to show vdr messages if they occur.
- Changed the manner the moving of the PiP window works. Now move mode
  is enabled or disabled by pressing the green button. In move mode,
  the cursor keys can be used to move the PiP window.
- Added some more predefined PiP window sizes. You have to reset your
  size in setup menu.
- Added support for ffmpeg 0.4.9 (should work with 0.4.8, too)
- Updated finnish translations (thanks to Rolf Ahrenberg)
- Cleaned up, dropped support for vdr versions older than 1.3.8


2005-02-03: Version 0.0.7.2
- adapted to VDR 1.3.19+ (should work with older versions, too)


2005-01-09: Version 0.0.7.1
- adapted to VDR 1.3.18+ (should work with older versions, too)


2004-10-26: Version 0.0.7
- adapted to VDR 1.3.13+ (should work with older versions, too)


2004-08-01: Version 0.0.6 (written by Andreas Regel)
- added channel swapping. It is now possible to swap the currently viewed
  channel with the pip channel by pressing the red key. (thanks to Sascha
  Volkenandt)
- added support for other aspect ratios than 4:3. Image is now correctly
  scaled and black border is added when necessary.
- added automatic frame dropping that always takes the last frame from
  the ringbuffer.
- added -D_GNU_SOURCE to DEFINES in Makefile
- fixed a memory leak: missing deletion of frame ringbuffer
- fixed a memory leak: missing frame deletion in case of a full ringbuffer
- small speedup through just putting relevant frames to ringbuffer
- updated finnish translations (thanks to Rolf Ahrenberg)


2004-06-13: Version 0.0.5 (written by Andreas Regel)
- added support for VDR version 1.3.7 and later (VDR 1.2.6 is still
  supported, no guarantee for 1.3.0-1.3.6)
- no patching required any longer with 1.3.7 and later
- fixed info window colors if palette patch is not applied (VDR up to 1.3.6,
  index patch must be applied)
- added german translations
- added finnish translations (thanks to Rolf Ahrenberg)


2004-02-04: Version 0.0.4 (written by Andreas Regel)

- a second osd window with some information about the viewed channel and
  event is now shown on screen


2004-01-24: Version 0.0.3 (written by Andreas Regel)

- new TS->ES remuxer: now using VDR's cRemux for TS->PES and some own code
  for PES->ES
- now using libavcodec from ffmpeg instead of mpeg2dec
- frames to decode configurable (I-frames, I-/P-frames, all frames)
- frame dropping configurable
- added new color depths:
 - 128 shades greyscale
 - 128 colors with variable palette using Wu's quantizer (patch needed)
- changed osd size setting to 6 configurable values


2004-01-03: Version 0.0.2

- Added setup menu
- OSD depth, size and position configurable (see README)
- OSD crop dimensions configurable
- OSD window size is calculated dynamically depending on stream dimensions
- Added translations


2004-01-02: Version 0.0.1

- Initial revision.


setup.c
/*
 * OSD Picture in Picture plugin for the Video Disk Recorder
 *
 * Copyright (C) 2004 - 2008 Andreas Regel <andreas.regel@powarman.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <vdr/config.h>

#include "setup.h"

const int kColorDepths = 5;
const int kSizes = 16;
const int kFrameModes = 3;
const int kFrameDrops = 4;
const int kShowInfoTypes = 4;
const int kInfoPositions = 4;

const char * ColorDepthItems[] = {NULL, NULL, NULL, NULL, NULL, NULL};   // initialized later
const char * ShowInfoItems[] = {NULL, NULL, NULL, NULL, NULL}; // initialized later
const char * InfoPositionItems[] = {NULL, NULL, NULL, NULL, NULL}; // initialized later
const char * FrameDropItems[] = {NULL, NULL, NULL, NULL, NULL}; // initialized later

const char * SizeItems[] = {
    "100x80",
    "120x96",
    "140x112",
    "160x128",
    "180x144",
    "200x160",
    "220x176",
    "240x192",
    "260x208",
    "280x224",
    "300x240",
    "450x360",
    "600x480",
    "320x180",
    "480x270",
    "640x360",
    NULL
};

const char * FrameModeItems[] = {
    "I",
    "I, P",
    "I, P, B"
};

cOsdPipSetup OsdPipSetup;

cOsdPipSetup::cOsdPipSetup(void)
{
    XPosition  = 50;
    YPosition  = 50;
    CropLeft   = 5;
    CropRight  = 5;
    CropTop    = 5;
    CropBottom = 5;
    ColorDepth = kDepthGrey16;
    Size = 5;
    FrameMode = kFrameModeI;
    FrameDrop = -1;
    SwapFfmpeg = 1;
    ShowInfo = 1;
    InfoWidth = 400;
    InfoPosition = kInfoBottomLeft;
}

bool cOsdPipSetup::SetupParse(const char *Name, const char *Value)
{
    if      (strcmp(Name, "XPosition")    == 0) XPosition    = atoi(Value);
    else if (strcmp(Name, "YPosition")    == 0) YPosition    = atoi(Value);
    else if (strcmp(Name, "CropLeft")     == 0) CropLeft     = atoi(Value);
    else if (strcmp(Name, "CropRight")    == 0) CropRight    = atoi(Value);
    else if (strcmp(Name, "CropTop")      == 0) CropTop      = atoi(Value);
    else if (strcmp(Name, "CropBottom")   == 0) CropBottom   = atoi(Value);
    else if (strcmp(Name, "ColorDepth")   == 0) ColorDepth   = atoi(Value);
    else if (strcmp(Name, "Size")         == 0) Size         = atoi(Value);
    else if (strcmp(Name, "FrameMode")    == 0) FrameMode    = atoi(Value);
    else if (strcmp(Name, "FrameDrop")    == 0) FrameDrop    = atoi(Value);
    else if (strcmp(Name, "SwapFfmpeg")   == 0) SwapFfmpeg   = atoi(Value);
    else if (strcmp(Name, "ShowInfo")     == 0) ShowInfo     = atoi(Value);
    else if (strcmp(Name, "InfoWidth")    == 0) InfoWidth    = atoi(Value);
    else if (strcmp(Name, "InfoPosition") == 0) InfoPosition = atoi(Value);
    else return false;
    return true;
}

cOsdPipSetupPage::cOsdPipSetupPage(void)
{
    m_NewOsdPipSetup = OsdPipSetup;
    m_NewOsdPipSetup.FrameDrop += 1;

    ColorDepthItems[0] = tr("Greyscale (16)");
    ColorDepthItems[1] = tr("Greyscale (256)");
    ColorDepthItems[2] = tr("Color (256, fixed)");
    ColorDepthItems[3] = tr("Color (128, variable)");
    ColorDepthItems[4] = tr("Color (256, dithered)");

    ShowInfoItems[0] = trVDR("no");
    ShowInfoItems[1] = tr("channel only");
    ShowInfoItems[2] = tr("simple");
    ShowInfoItems[3] = tr("complete");

    InfoPositionItems[0] = tr("top left");
    InfoPositionItems[1] = tr("top right");
    InfoPositionItems[2] = tr("bottom left");
    InfoPositionItems[3] = tr("bottom right");

    FrameDropItems[0] = tr("automatic");
    FrameDropItems[1] = tr("none");
    FrameDropItems[2] = tr("1 frame");
    FrameDropItems[3] = tr("2 frames");

    Add(new cMenuEditIntItem(tr("X Position"), &m_NewOsdPipSetup.XPosition, 0, 600));
    Add(new cMenuEditIntItem(tr("Y Position"), &m_NewOsdPipSetup.YPosition, 0, 470));
    Add(new cMenuEditIntItem(tr("Crop left"), &m_NewOsdPipSetup.CropLeft, 0, 80));
    Add(new cMenuEditIntItem(tr("Crop right"), &m_NewOsdPipSetup.CropRight, 0, 80));
    Add(new cMenuEditIntItem(tr("Crop at top"), &m_NewOsdPipSetup.CropTop, 0, 80));
    Add(new cMenuEditIntItem(tr("Crop at bottom"), &m_NewOsdPipSetup.CropBottom, 0, 80));
    Add(new cMenuEditStraItem(tr("Color depth"), &m_NewOsdPipSetup.ColorDepth, kColorDepths, ColorDepthItems));
    Add(new cMenuEditStraItem(tr("Size"), &m_NewOsdPipSetup.Size, kSizes, SizeItems));
    Add(new cMenuEditStraItem(tr("Frames to display"), &m_NewOsdPipSetup.FrameMode, kFrameModes, FrameModeItems));
    Add(new cMenuEditStraItem(tr("Drop frames"), &m_NewOsdPipSetup.FrameDrop, kFrameDrops, FrameDropItems));
    Add(new cMenuEditBoolItem(tr("Swap FFMPEG output"), &m_NewOsdPipSetup.SwapFfmpeg));
    Add(new cMenuEditStraItem(tr("Show info window"), &m_NewOsdPipSetup.ShowInfo, kShowInfoTypes, ShowInfoItems));
    Add(new cMenuEditIntItem(tr("Info window width"), &m_NewOsdPipSetup.InfoWidth, 200, 600));
    Add(new cMenuEditStraItem(tr("Info window position"), &m_NewOsdPipSetup.InfoPosition, kInfoPositions, InfoPositionItems));
}

cOsdPipSetupPage::~cOsdPipSetupPage()
{
}

void cOsdPipSetupPage::Store(void)
{
    OsdPipSetup = m_NewOsdPipSetup;
    OsdPipSetup.FrameDrop -= 1;

    SetupStore("XPosition", OsdPipSetup.XPosition);
    SetupStore("YPosition", OsdPipSetup.YPosition);
    SetupStore("CropLeft", OsdPipSetup.CropLeft);
    SetupStore("CropRight", OsdPipSetup.CropRight);
    SetupStore("CropTop", OsdPipSetup.CropTop);
    SetupStore("CropBottom", OsdPipSetup.CropBottom);
    SetupStore("ColorDepth", OsdPipSetup.ColorDepth);
    SetupStore("Size", OsdPipSetup.Size);
    SetupStore("FrameMode", OsdPipSetup.FrameMode);
    SetupStore("FrameDrop", OsdPipSetup.FrameDrop);
    SetupStore("SwapFfmpeg", OsdPipSetup.SwapFfmpeg);
    SetupStore("ShowInfo", OsdPipSetup.ShowInfo);
    SetupStore("InfoWidth", OsdPipSetup.InfoWidth);
    SetupStore("InfoPosition", OsdPipSetup.InfoPosition);
}

